src/Estimators/WalkerTraceManager.h

```diff
@@ -0,0 +1,465 @@
+    }
+    app_log() << pad << "end WalkerTraceBuffer(" << label << ")" << std::endl;
+  }
+
+  inline void register_hdf_data(hdf_archive& f)
+  {
+    auto& top = label;
+    f.push(top);
+    f.push("data_layout");
+    for(auto& wqi: quantity_info)
+    {
+      f.push(wqi.name);
+      f.write(wqi.dimension,    "dimension"  );
+      f.write(wqi.shape,        "shape"      );
+      f.write(wqi.size,         "size"       );
+      f.write(wqi.unit_size,    "unit_size"  );
+      f.write(wqi.buffer_start, "index_start");
+      f.write(wqi.buffer_end,   "index_end"  );
+      f.pop();
+    }
+    f.pop();
+    f.pop();
+    if (!f.open_groups())
+      throw std::runtime_error("WalkerTraceBuffer(" + label + ")::register_hdf_data() some hdf groups are still open at the end of registration");
+    hdf_file_pointer = 0;
+  }
+
+
+  inline void write_hdf(hdf_archive& f)
+  {
+    write_hdf(f,hdf_file_pointer);
+  }
+
+
+  inline void write_hdf(hdf_archive& f, hsize_t& file_pointer)
+  {
+    //if (verbose) app_log()<<"WalkerTraceBuffer("<<label<<")::write_hdf "<<file_pointer<<"  "<<buffer.size(0)<<" "<<buffer.size(1)<<std::endl;
+    auto& top = label;
+    dims[0] = buffer.size(0);
+    dims[1] = buffer.size(1);
+    if (dims[0] > 0)
+    {
+      f.push(top);
+      h5d_append(f.top(), "data", file_pointer, buffer.dim(), dims, buffer.data());
+      f.pop();
+    }
+    f.flush();
+  }
+};
+
+
+
+struct WalkerTraceState 
+{
+  bool traces_active;
+  int step_period;
+  bool verbose;
+
+  WalkerTraceState()
+  {
+    reset_permissions();
+    step_period = 1;
+  }
+
+  inline void reset_permissions()
+  {
+    traces_active = false;
+    verbose       = false;
+  }
+};
+
+
+
+template<class QT, class PT> class Walker;
+class ParticleSet;
+class TrialWaveFunction;
+class QMCHamiltonian;
+using MCPWalker = Walker<QMCTraits, PtclOnLatticeTraits>;
+
+
+
+class WalkerTraceCollector
+{
+public:
+  std::vector<size_t>     steps;
+  std::vector<WTraceReal> energies;
+  WalkerTraceBuffer<WTraceInt>  walker_property_int_buffer;
+  WalkerTraceBuffer<WTraceReal> walker_property_real_buffer;
+  WalkerTraceBuffer<WTraceReal> walker_particle_real_buffer;
+
+  std::unordered_set<std::string> properties_include;
+  std::vector<size_t>             property_indices;
+  int energy_index;
+
+private:
+  WalkerTraceState state;
+  
+  // temporary (contiguous) storage for awful ParticleAttrib<> quantities
+  Array<WTraceReal  , 2> Rtmp;
+  Array<WTraceReal  , 1> Stmp;
+  Array<WTracePsiVal, 2> Gtmp;
+  Array<WTracePsiVal, 1> Ltmp;
+
+public:
+  WalkerTraceCollector();
+
+  void set_state(const WalkerTraceState& tms);
+
+  void startBlock();
+
+  void collect(MCPWalker& walker, ParticleSet& pset, TrialWaveFunction& wfn, QMCHamiltonian& ham, int step=-1);
+
+  void check_buffers();
+
+private:
+  void reset_buffers();
+};
+
+
+
+class WalkerTraceManager
```

Comment:
This is the third full class in the source file. Please make `WalkerTraceBuffer` and `WalkerTraceCollector` into separate files and properly unit tested.

Reply:
Done